src/corecel/data/LdgIterator.hh

```diff
@@ -11,38 +11,67 @@
 #include <type_traits>
 
 #include "corecel/Macros.hh"
-#include "corecel/Types.hh"
 #include "corecel/cont/Span.hh"
-#include "corecel/math/Algorithms.hh"
 
-#include "detail/LdgIteratorImpl.hh"
+#include "LdgTraits.hh"
 
 namespace celeritas
 {
+//---------------------------------------------------------------------------//
+//! Whether a type is supported by __ldg
+template<class T>
+inline constexpr bool is_ldg_supported_v
+    = !std::is_void_v<typename LdgTraits<T>::underlying_type>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Wrap the low-level CUDA/HIP "load read-only global memory" function.
+ *
+ * This low-level capability allows improved caching because we're \em
+ * promising that the data is not mem. For CUDA the load is cached
+ *
+ * \warning The target address must be read-only for the lifetime of the
+ * kernel. This is generally true for Params data but not State data.
+ */
+template<class T>
+CELER_CONSTEXPR_FUNCTION T ldg(T const* ptr)
+{
+    using TraitsT = LdgTraits<T>;
+    using underlying_type = typename TraitsT::underlying_type;
+    static_assert(std::is_arithmetic_v<underlying_type>,
+                  "Only types with arithmetic underlying types are supported "
+                  "by __ldg");
+
+    auto const* data_ptr = TraitsT::data(ptr);
+#if CELER_DEVICE_COMPILE
+    return T{__ldg(data_ptr)};
+#else
+    return T{*data_ptr};
+#endif
+}
+
 //---------------------------------------------------------------------------//
 /*!
  * Iterator for read-only device data in global memory.
+ * \tparam T value type being accessed
  *
  * This wraps pointer accesses with the \c __ldg intrinsic to load
  * read-only data using texture cache.
  */
 template<class T>
 class LdgIterator
 {
-    static_assert(detail::is_ldg_supported_v<T>,
-                  "LdgIterator requires const arithmetic, OpaqueId or "
-                  "enum type");
-
-  private:
-    using LoadPolicyT = detail::LdgLoader<T>;
+    using TraitsT = LdgTraits<T>;
```

Comment:
Should we keep a 

```c++
static_assert(typename LdgTraits<T>::underlying_tyoe != void);
```

here to make the compiler error more easy to debug in case of misuse?